cdap-ui/app/hydrator/controllers/create/toppanel-ctrl.js

```diff
@@ -481,26 +481,24 @@ class HydratorPlusPlusTopPanelCtrl {
   }
 
   startOrStopPreview() {
-    if (this.doesPreviewHaveEmptyMacros) {
-      this.doStartOrStopPreview();
-    } else {
-      // Validate and show runtime arguments if there are
-      // un-fulfilled macros.
-      this.toggleConfig();
-    }
-  }
-
-  doStartOrStopPreview() {
-    this.getRuntimeArguments()
-      .then(() => {
+    if (this.previewRunning) {
+      this.stopPreview();
+    } else if (this.doesPreviewHaveEmptyMacros) {
+      this.getRuntimeArguments().then(() => {
         if (this.previewRunning) {
           this.stopPreview();
```

Comment:
we won't ever get to this condition, can we?

Reply:
If getRuntimeArguments call takes a while to finish, the preview state could change and we might hit this edge case I think. I doubt it's likely to happen in a real scenario but just to be safe, I did not remove this. Let me know if I should remove this, will make the change.

Reply:
in that case i think the right thing to do is to do nothing. When the getRuntimeArguments modal open, the expectation is for the preview to start. If it's already running, it shouldn't stop the currently running preview.